### PR TITLE
feat(bindings): best-effort serialize circular structures

### PR DESCRIPTION
**Pull request by pavelfeldman**

### CODE REVIEW

File: packages/playwright-core/src/server/isomorphic/utilityScriptSerializers.ts [hunk 1]
```diff
@@ -23,129 +23,142 @@ export type SerializedValue =
...
 
-function isRegExp(obj: any): obj is RegExp {
-  return obj instanceof RegExp || Object.prototype.toString.call(obj) === '[object RegExp]';
-}
+export type HandleOrValue = { h: number } | { fallThrough: any };
 
-function isDate(obj: any): obj is Date {
-  return obj instanceof Date || Object.prototype.toString.call(obj) === '[object Date]';
-}
+export function source(ignoreCycles: boolean = false) {
```
Comment (yury-s) [inline packages/playwright-core/src/server/isomorphic/utilityScriptSerializers.ts:28]: source -> `createSerializers()` ?

File: packages/playwright-core/src/server/page.ts [hunk 2]
```diff
@@ -718,12 +726,12 @@ export class PageBinding {
   constructor(name: string, playwrightFunction: frames.FunctionWithSource, needsHandle: boolean) {
     this.name = name;
     this.playwrightFunction = playwrightFunction;
-    this.source = `(${addPageBinding.toString()})(${JSON.stringify(name)}, ${needsHandle})`;
+    this.source = `(${addPageBinding.toString()})(${JSON.stringify(name)}, ${needsHandle}, (${source})(true))`;
```
Comment (yury-s) [inline packages/playwright-core/src/server/page.ts:729]: Why can't we send `(${source})(true)` once into the page?

**Review by yury-s — APPROVED** [review]

File: packages/playwright-core/src/server/isomorphic/utilityScriptSerializers.ts [hunk 3]
```diff
@@ -23,129 +23,142 @@ export type SerializedValue =
...
+  function serialize(value: any, handleSerializer: (value: any) => HandleOrValue, visited: Set<any>): SerializedValue {
+    const result = handleSerializer(value);
+    if ('fallThrough' in result)
+      value = result.fallThrough;
+    else
       return result;
+
+    if (visited.has(value)) {
+      if (ignoreCycles)
+        return '[Circular Ref]';
```
Comment (dgozman) [inline packages/playwright-core/src/server/isomorphic/utilityScriptSerializers.ts:92]: - I am somewhat worried about performance, now that you can pass whatever to the binding.
- On the other hand, we can go further and restore cycles like we do with handles.

File: packages/playwright-core/src/server/isomorphic/utilityScriptSerializers.ts (hunk 1 above)
Comment (dgozman) [reply packages/playwright-core/src/server/isomorphic/utilityScriptSerializers.ts:28]: Yeah, things like `typeof source` look suspicious :)

**Review by dgozman — APPROVED** [review]

File: packages/playwright-core/src/server/isomorphic/utilityScriptSerializers.ts (hunk 3 above)
Comment (pavelfeldman) [reply packages/playwright-core/src/server/isomorphic/utilityScriptSerializers.ts]: Aliasing Window / Document / Node

File: packages/playwright-core/src/server/page.ts (hunk 2 above)
Comment (pavelfeldman) [reply packages/playwright-core/src/server/page.ts:732]: Same applies to addPageBinding.

**Pull request merged**